plugins/module_utils/openssh/utils.py

```diff
@@ -33,3 +66,309 @@ def parse_openssh_version(version_string):
+
+
+# See https://datatracker.ietf.org/doc/html/rfc4251#section-5 for SSH data types
+class OpensshParser(object):
+    """Parser for OpenSSH encoded objects"""
+    BOOLEAN_OFFSET = 1
+    UINT32_OFFSET = 4
+    UINT64_OFFSET = 8
+
+    def __init__(self, data):
+        if not isinstance(data, (bytes, bytearray)):
+            raise TypeError("Data must be bytes-like not %s" % type(data))
+
+        self._data = memoryview(data) if PY3 else data
+        self._pos = 0
+
+    def boolean(self):
+        self._check_position(self.BOOLEAN_OFFSET)
+
+        next_pos = self._pos + self.BOOLEAN_OFFSET
+        value = _BOOLEAN.unpack(self._data[self._pos:next_pos])[0]
+        self._pos = next_pos
+        return value
+
+    def uint32(self):
+        self._check_position(self.UINT32_OFFSET)
+
+        next_pos = self._pos + self.UINT32_OFFSET
+        value = _UINT32.unpack(self._data[self._pos:next_pos])[0]
+        self._pos = next_pos
+        return value
+
+    def uint64(self):
+        self._check_position(self.UINT64_OFFSET)
+
+        next_pos = self._pos + self.UINT64_OFFSET
+        value = _UINT64.unpack(self._data[self._pos:next_pos])[0]
+        self._pos = next_pos
+        return value
+
+    def string(self):
+        length = self.uint32()
+        self._check_position(length)
+
+        next_pos = self._pos + length
+        value = self._data[self._pos:next_pos]
+        self._pos = next_pos
+        # Cast to bytes is required as a memoryview slice is itself a memoryview
+        return value if not PY3 else bytes(value)
+
+    def mpint(self):
+        return self._big_int(self.string(), "big", signed=True)
+
+    def name_list(self):
+        raw_string = self.string()
+        return raw_string.decode('ASCII').split(',')
+
+    # Convenience function, but not an official data type from SSH
+    def string_list(self):
+        result = []
+        raw_string = self.string()
+
+        if raw_string:
+            parser = OpensshParser(raw_string)
+            while parser.remaining_bytes():
+                result.append(parser.string())
+
+        return result
+
+    # Convenience function, but not an official data type from SSH
+    def option_list(self):
+        result = []
+        raw_string = self.string()
+
+        if raw_string:
+            parser = OpensshParser(raw_string)
+
+            while parser.remaining_bytes():
+                name = parser.string()
+                data = parser.string()
+                if data:
+                    # data is doubly-encoded
+                    data = OpensshParser(data).string()
+                result.append((name, data))
+
+        return result
+
+    def seek(self, offset):
+        self._check_position(offset)
+        self._pos = self._pos + offset
+
+        return self._pos
+
+    def remaining_bytes(self):
+        return len(self._data) - self._pos
+
+    def _check_position(self, offset):
+        if self._pos + offset > len(self._data):
```

Comment:
How about first checking that `offset >= 0`? Negative values indicate that something is broken, and also then the `elif` below cannot happen.

Reply:
Internally within the parser methods the buffer should only move forward, but `seek` is exposed to manually manipulate the cursor forwards and backwards which is why `_check_position` permits negative offsets.

Reply:
In that case, all the code locations that use `_check_position` and pass something read by the parser have to make sure that the passed value is non-negative. (Ok, I just noticed that the only place where that happens reads the data with `uint32()`, which is never negative.)